docs: DOC-612: Engine internals (#8190)
deephaven/deephaven-core: docs/groovy/conceptual/deephaven-design.md

| ---------------------------------------- | ------------- | ------------------------ |
| **Filtering** (`where`)                  | New (subset)  | Shared                   |
| **Column derivation** (`update`, `view`) | Shared        | Mixed (shared + new)     |
| **Sorting** (`sort`)                     | New (flat)    | Shared (via redirection) |
| **Joining** (`naturalJoin`, etc.)        | New or shared | Mixed                    |
| **Aggregation** (`aggBy`, etc.)          | New           | New                      |

This sharing model, combined with [incremental updates](./table-update-model.md) through the [DAG](./dag.md), enables Deephaven to handle complex queries on large, rapidly-changing datasets efficiently.

### How operations stay live

The listener attachment in step 5 of each operation is what makes Deephaven tables "live." When a parent table updates:

1. The parent's `notifyListeners` method enqueues update notifications for all child listeners
2. Each listener receives a `TableUpdate` describing which rows were added, removed, modified, or shifted, along with information about which columns changed

> 3. The listener processes the update and propagates its own update downstream. Most operations process only the changed rows, though some (like certain filters) may need to re-examine additional rows

Reviewer: periods?

This continues through the entire DAG. A single source change cascades through filters, joins, and aggregations — with most operations processing only the delta, not the full dataset. The engine processes this DAG within a single update cycle. The default cycle targets 1000ms intervals (including both processing and idle time), though individual cycles may take longer if the workload requires it. The results then flow asynchronously through the [API layer](#the-live-data-stack) via Barrage to connected clients and UI components.

> This is the technical foundation of [Live Dataframes](#the-live-data-stack): a table is static (if its source never changes) or live (if connected to streaming data), and most downstream operations preserve that behavior. (Some operations, like [`snapshot`](../reference/table-operations/snapshot/snapshot.md), intentionally produce static results from live sources.)

Reviewer: ```suggestion
This is the technical foundation of [Live Dataframes](#the-live-data-stack): a table is static (if its source never changes) or live (if connected to streaming data), and all downstream operations preserve that behavior (with the exception of operations like [`snapshot`](../reference/table-operations/snapshot/snapshot.md) that intentionally produce static results from live sources.)
```

all/ almost all?

- **[`web-client-ui`](https://github.com/deephaven/web-client-ui)**: JavaScript/TypeScript components for building custom web applications with full control over the front-end.

## gRPC APIs for polyglot interoperability

> Deephaven’s core API is implemented using polyglot technologies that allow for compatible client (or server!) implementations in almost any language. It is composed of several complementary modules, but its Arrow Flight service and Table service are foremost. These offer high-performance data transport -- specifically organized to include real-time and updating data -- and a table manipulation API that mirrors the Deephaven engine’s internal compute paradigm. You can read more about our API itself [here](./deephaven-core-api.md).

Reviewer: ```suggestion
Deephaven’s core API is implemented using polyglot technologies that allow for compatible client (or server!) implementations in almost any language. It is composed of several complementary modules, but its Arrow Flight service and Table service are foremost. These offer high-performance data transport - specifically organized to include real-time and updating data - and a table manipulation API that mirrors the Deephaven engine’s internal compute paradigm. You can read more about our API itself [here](./deephaven-core-api.md).
```

## Distributing DAGs and global consistency

> At Deephaven, we believe that our approach to propagating static and updating tabular data will revolutionize distributed data systems development. It represents a powerful new model.

Reviewer: The second sentence in this sentence doesn't seem necessary.
